src/PrettyPrinting.jl

```diff
@@ -1591,23 +1591,81 @@ function _write_line(io::IOCustom, str::AbstractString)
     spaceleft = c - ind - io.printed
   end
   #@show spaceleft
-  firstlen = min(spaceleft, length(str))
-  firststr = str[1:firstlen]
+  # The following code deals with line wrapping of Unicode text, including
+  # double-width symbols and more.
+  _graphemes = Base.Unicode.graphemes(str)
+  firstlen = min(spaceleft, length(_graphemes))
+  # make an iterator over valid indices
   if io.lowercasefirst
-    written += write(io.io, lowercasefirst(firststr))
+    firstiter = Base.Iterators.take(_graphemes, firstlen)
+    firstiter = join(firstiter)
+    firstiter = Base.Unicode.graphemes(lowercasefirst(firstiter))
     io.lowercasefirst = false
   else
-    written += write(io.io, firststr)
+    firstiter = Base.Iterators.take(_graphemes, firstlen)
     io.lowercasefirst = false
   end
-  io.printed += textwidth(firststr)
-  reststr = str[firstlen + 1:end]
-  it = Iterators.partition(1:textwidth(reststr), c - ind > 0 ? c - ind : c)
-  for i in it
+  restiter = Base.Iterators.drop(_graphemes, firstlen)
+  firststr = join(firstiter)
+  if length(firstiter) == textwidth(firststr)
+    written += write(io.io, firststr)
+    io.printed += textwidth(firststr)
+  else
+    #firstline is wider than number of graphemes
+    partcollect = collect(firstiter)
+    printstr = ""
+    j = 1
+    while textwidth(printstr) < (c - ind > 0 ? c - ind : c)
+      printstr *= partcollect[j]
+      j += 1
+      if j > length(partcollect)
+        break
+      end
+    end
+    written += write(io.io, printstr)
+    io.printed += textwidth(printstr)
+
+    #the spillover string
     written += write(io.io, "\n")
     written += write_indent(io)
-    written += write(io.io, reststr[i])
-    io.printed = textwidth(reststr[i])
+    printstr = join(collect(firstiter)[j:end])
+    written += write(io.io, printstr)
+    io.printed += textwidth(printstr)
+  end
+  it = Iterators.partition(1:length(restiter), c - ind > 0 ? c - ind : c)
+  restcollect = collect(restiter)
+  for i in it
+    # partitions of the spillover text
+    partcollect = restcollect[i]
+    partstr = join(partcollect)
+    if (textwidth(partstr) < (c - ind > 0 ? c - ind : c)) || length(i) == textwidth(partstr)
+      written += write(io.io, "\n")
+      written += write_indent(io)
+      written += write(io.io, partstr)
+      io.printed = textwidth(partstr)
```

Comment:
```suggestion
    width = textwidth(partstr)
    if (width < (c - ind > 0 ? c - ind : c)) || length(i) == width
      written += write(io.io, "\n")
      written += write_indent(io)
      written += write(io.io, partstr)
      io.printed = width
```
or with `limit` available as suggested:
```suggestion
    width = textwidth(partstr)
    if width < limit || length(i) == width
      written += write(io.io, "\n")
      written += write_indent(io)
      written += write(io.io, partstr)
      io.printed = width
```